[Merged by Bors] - Speed up ATX queries

ivan4th opened the pull request:
## Motivation

Lots of SQL queries to the state db take long time to execute, e.g. cache warmup is taking a lot of time

## Description

This PR splits ATX blobs into a separate table (`atx_blobs`).
It also adds `nonce` column to each row in the `atxs` table, improving the query performance.
The database upgrade procedure had to be modified so that each migration is done in a separate transaction, and vacuuming can be done between migrations if requested by the `db-vacuum-state` setting.  This may seemingly cause "partial" upgrades when switching to a newer go-spacemesh version, rendering the database unusable with older versions, but given that the local DB is updated separately, the data directory can't be expected to be usable with an older version after such an attempt anyway.

Timings on an M3 Max Mac:
* cache warmup before upgrade on v1.4.4: up to 6m
* db upgrade
  * splitting ATX blobs 27m
  * vacuuming 14m
  * updating nonces in the `atxs` table 11m
  * total ~52m
* cache warmup after upgrade: 15s

Fixes #5701

## Test Plan

Verified on a mainnet node.

## TODO
- [X] initial speedup of cache warmup via subquery
- [X] split ATX blobs from atxs table
- [X] store nonce in each atxs row
- [X] do db upgrade steps as separate transactions
- [X] do not publish nonces as part of ATX when not necessary
- [X] add previous ATX ID field to atxs table
- [X] update CHANGELOG.md (after PR update)

ivan4th committed:
Speed up cache warmup

ivan4th committed:
Update CHANGELOG.md

ivan4th commented:
bors try

dshulyak commented on sql/atxs/atxs.go:
```diff
@@ -569,27 +571,53 @@ func IterateAtxsData(
+		     where a1.pubkey = a.pubkey and a1.epoch < a.epoch and a1.nonce is not null
+		     order by a1.epoch desc
+		     limit 1
+		   )) as nonce,
```
it is also possible to copy nonce from the previous atx on insertion, it will add very little data but completely avoid this subqueries, and also this bug https://github.com/spacemeshos/go-spacemesh/issues/5701

dshulyak commented on sql/atxs/atxs.go:
```diff
@@ -569,27 +571,53 @@ func IterateAtxsData(
+		// func(stmt *sql.Statement) {
+		// 	stmt.BindInt64(1, int64(from.Uint32()))
+		// 	stmt.BindInt64(2, int64(to.Uint32()))
+		// },
```
i guess it means that indexes are not useful for this query at all, and reading whole table is faster.
i tried to decouple blobs from the rest sometime ago https://github.com/spacemeshos/go-spacemesh/pull/5670

ivan4th replied:
Although splitting the blobs did improve performance of multiple ATX queries substantially, in this particular case it's still more efficient not to use this extra `where` condition

fasmat commented:
What is the difference between `IterateAtxsData` and `IterateAtxsOps`? They seem to be targeting the same purpose.

dshulyak commented:
the second is used in api and allows a bit of programmability, the first one is used in warmup.go (which is a subjet of this change)

fasmat commented:
There is also `IterateForGrading` which looks similar to the other two. Would it make sense to only have `IterateAtxsOps` and pass the `builder.Operations` needed for the specific query?

I know `IterateAtxsOps` decodes the ATXs from their binary representation, but that's gonna change with ATX v2, where aim to not need to do that any more.

cc @poszu

dshulyak commented:
unlike the other two IterateForGrading selects unique piece of information that is not used in those other two. it would be wasteful to always do a join and select that piece of information in other queries. also in this change there is another part that selects nonce which wouldn't make sense to use in IterateForGrading, and it certainly would be problematic to add it to IterateOps.

it doesn't sound like a good idea to break down atx into multiple fields and store all of them separately. maybe you can try to keep blobs as they are. and save subset of fields that are relevant for consensus from atx with version 2, in a way similar to current schema.

blob is stored in original form so that it can be fetched in sync without re-encoding. if you break down atx into multiple fields it will make it less efficient.

the rest of the fields are decoupled and used only for consensus and sometimes for indexing. which is still imperfect and i think better solution would be something like in https://github.com/spacemeshos/go-spacemesh/pull/5670

fasmat commented:
@dshulyak the goal is to keep only the minimum necessary information in individual columns in the table.

The main problem with relying on decoding ATXs when fetching them from the DB is, that we then need to put logic into the SQL package to be able to decode different blobs of ATXs based on their publish epoch. In my opinion this should be the sole responsibility of the `activation.Handler` and other parts of the codebase should ideally not even be aware about different encoding schemas of ATXs.

For gossip / fetch the handling functions don't need to be aware how the blob data encodes an ATX but just serve / gossip it if needed. As fallback one can use the ATX ID to fetch the binary blob and decode the ATX themselves if needed somewhere, with the caveat that they then need to be aware of possible different encoding schemas of ATXs.

Right now we have multiple ways of representing the same data structure:

- `types.ActivationTx`
- `types.VerifiedActivationTx`
- `types.ActivationTxHeader`
- `atxsdata.ATX`

With some of them performing double duties and - at least for the first 3 - tight coupling between them. For me ideally we only had 2 types:

- `types.ATX`
- `activation.ATXDto` (actually one for every version of ATX encoding scheme we use now or in the future)

Where the first contains all the information relevant for consensus and in the best case maps fields 1:1 to the columns in the atx table in the DB and the second is the representation of the ATX as it is sent over the wire and should ONLY be used for that. The first can be used anywhere and be freely extended or changed without having to fear to break backwards compatibility. Queries would always return an `types.ATX` (with some fields possibly being `nil` or zero if that speeds up specific queries) and only the `activation` package is aware of the actual binary representation of an ATX so it can decode them in the `Handler` and encode them in the `Builder`. Possibly `checkpoint` needs to be aware as well - because it cares about `PrevATXID` and `PositioningATX` that are not available as columns in the `atx` table - besides that it can just (re)store the ATXs as bytes and not care about decoding them.

I'm only advocating for a leaner surface to the `sql/atxs` package. Ideally there would only be one `Iterate` method that is flexible enough to handle the 3 use cases we already have and not need 3 individual functions for it. `builder.Options` could be extended with `join`s and with `select`s.

I've seen #5670 but I don't fully understand how this improves performance? Why does it matter if the blob data is in the same or a different table? I'm not advocating against it. It could even help keeping the separation as suggested clear in the code. Would moving it to a different table not cause functions like `IterateAtxsOps`, `Get` and `GetByEpochAndNodeID` to break (also indirectly anything that uses `types.ActivationTxHeader`)?

ivan4th commented:
Probably simplifying somewhat, but #5670 should improve SQLite performance b/c with blobs taken out of `atxs` table, the whole table should basically fit in the cache (RAM) and thus all the operations involving a subset of ATXs but not actual blobs should be faster

ivan4th committed:
sql: use a separate table for ATX blobs

ivan4th committed:
Always include nonce in ATXs

ivan4th committed:
sql: commit after each successful migration

Also, vacuum the database as soon as needed (possibly between migrations)

fasmat commented on activation/activation.go:
```diff
@@ -652,24 +652,12 @@ func (b *Builder) createAtx(
-			b.log.Warn("failed to get VRF nonce for ATX", zap.Error(err), log.ZShortStringer("smesherID", sig.NodeID()))
-			break
-		}
-		if nipostState.VRFNonce != oldNonce {
-			nonce = &nipostState.VRFNonce
-		}
 	}
-
+	nonce := &nipostState.VRFNonce
```
Why always include the VRFNonce? It only needs to be included in the initial ATX and in case it changes (i.e. when a user resizes their PoST).

ivan4th replied:
Always including nonce in ATX substantially speeds up lots of queries, see this comment https://github.com/spacemeshos/go-spacemesh/pull/5791#discussion_r1545473980
and the issue https://github.com/spacemeshos/go-spacemesh/issues/5701

fasmat replied:
No, sorry. There is no point in broadcasting values to other peers that they already know about. I don't have an issue with storing the VRFNonce for every ATX in the DB to speed up DB queries, but I object to including it in every broadcast ATX.

ivan4th replied:
Reverted this change

fasmat commented on atxsdata/warmup_test.go:
```diff
@@ -74,14 +74,6 @@ func TestWarmup(t *testing.T) {
 		require.Error(t, err)
 		require.Nil(t, c)
 	})
-	t.Run("missing nonce", func(t *testing.T) {
-		db := sql.InMemory()
-		data := gatx(types.ATXID{1, 1}, 1, types.NodeID{1}, nil)
-		require.NoError(t, atxs.Add(db, &data))
-		c, err := Warm(db, 1)
-		require.Error(t, err)
-		require.Nil(t, c)
-	})
```
The majority of existing ATXs do not have a VRFNonce in them

ivan4th replied:
Restored the test, it was removed by mistake

fasmat commented:
It makes no sense to include the VRFNonce in every published ATX. For any given identity it usually never changes. If we wouldn't support PoST resizing we would only ever need to include it in the initial ATX (similar to CommitmentATX). But since an existing nonce can become invalid if the user chooses to increase their PoST size we must have an option to update the VRFNonce of an identity.

If you want to simplify the lookup of the nonce instead we could make the `nonce` column `NOT NULL` and set it to the same value as in the previous ATX of the same NodeID if it is not set in the broadcast ATX. I would however also avoid this if possible. Adding the VRFNonce to all existing rows is a lot of redundant data, 8 bytes per ATX, so in the worst case that would be 80 MB of redundant data that is stored and could just be looked up with an existing DB query. Additionally it would be ~ 20 MB of data that needs to be broadcast every epoch - of data that every node already has?

fasmat commented on activation/activation_test.go:
```diff
@@ -1024,7 +1024,8 @@ func TestBuilder_PublishActivationTx_PrevATXWithoutPrevATX(t *testing.T) {
 			r.Nil(atx.CommitmentATX)
-			r.Nil(atx.VRFNonce)
+			r.NotNil(atx.VRFNonce)
+			r.Equal(nonce, *atx.VRFNonce)
```
The vast majority of non-initial ATXs have no `VRFNonce`.

ivan4th replied:
Reverted the change

ivan4th commented:
@fasmat ok I'll adjust the patch not to broadcast the nonce, yet we need to include it in the `atxs` rows.
Even if it's 80Mb increase it's close to nothing compared to the database size (up to ~26 GiB currently) and it speeds up a lot of queries substantially, e.g. warming up the cache is down to ~15s from ~1m 26s on my laptop compared to what can be achieved using a subquery

fasmat commented:
> @fasmat ok I'll adjust the patch not to broadcast the nonce, yet we need to include it in the `atxs` rows. Even if it's 80Mb increase it's close to nothing compared to the database size (up to ~26 GiB currently) and it speeds up a lot of queries substantially, e.g. warming up the cache is down to ~15s from ~1m 26s on my laptop compared to what can be achieved using a subquery

I see no issue in adding the value to the DB 🙂, I just want to avoid having to broadcast the nonce with every ATX.

ivan4th committed:
activation: revert always including nonce in published ATXs

ivan4th committed:
Merge branch 'develop' into fix/quicker-cache-warmup

ivan4th committed:
Speed up ATX blob split migration

ivan4th committed:
sql: don't modify ATX blob when adding nonces to the db

ivan4th committed:
sql: populate prev_id field in the state migration 17

ivan4th committed:
Fix nonce handling in activation and checkpoint pkgs

ivan4th committed:
Don't use node+epoch VRF query in the ATX handler

ivan4th committed:
Fix lint issue

ivan4th committed:
Fix CHANGELOG.md

ivan4th committed:
atxsdata: restore missing nonce test

ivan4th commented:
* reverted the change which was including nonces in each published ATX
* added `prev_id` field to the `atxs` table
* instead `VRFNonce` function (which queries nonce by epoch + node id) used `NonceByID` (which queries `atxs.nonce` field which is now always populated) where possible
* made the migration somewhat faster by means of re-creating the `atxs` table, and also by creating the indices after populating the `atxs` and `atx_blobs` tables. This comes at the price of not having `PRIMARY KEY` column in both tables as SQLite can't add PK to a table after the table has already been created, but the unique indices serve the same purpose and have no differences in performance

ivan4th committed:
Fix param naming

poszu commented on sql/migrations/state_0017_migration.go:
```diff
@@ -0,0 +1,366 @@
+package migrations
```
There should probably be a `0017_xyz.sql` empty migration file (with a comment) to avoid somebody creating it by mistake.

ivan4th replied:
Comments in the scripts appear to cause our database code to crash (probably will need to fix this separately), so added an empty placeholder file

fasmat replied:
We do have comments in `.sql` files, see for instance here: https://github.com/spacemeshos/go-spacemesh/blob/3fb7bc580b99a972deae42161b588018cad053d5/sql/migrations/local/0002_extend_initial_post.sql#L1-L2

poszu commented on sql/migrations/state/0016_atx_blob.sql:
```diff
@@ -0,0 +1,46 @@
+    epoch               INT NOT NULL,
+    effective_num_units INT NOT NULL,
+    commitment_atx      CHAR(32),
+    nonce               UNSIGNED LONG INT,
```
Why not `NOT NULL`? You could set the default to 0 here and fix it in the 17th in-code migration. Other fields could probably also be `NOT NULL`. Wdyt?

ivan4th replied:
As said in other comments, I so far failed to set `NOT NULL` here b/c it requires a lot of modifications in multiple tests across `go-spacemesh` which don't care about the nonces and ATX chains. I started doing it but it was taking a bit too much time, so for now I thought we better postpone that for a follow-up

ivan4th replied:
Noted the problem as tech debt in #5844

poszu commented on sql/migrations/state/0016_atx_blob.sql:
```diff
@@ -0,0 +1,46 @@
+    pubkey              CHAR(32),
+    coinbase            CHAR(24),
+    received            INT NOT NULL
+, validity INTEGER DEFAULT false);
```
```suggestion
    received            INT NOT NULL,
    validity INTEGER DEFAULT false
);
```

ivan4th replied:
Fixed

fasmat commented on sql/atxs/atxs.go:
```diff
@@ -350,8 +359,66 @@ func getBlob(ctx context.Context, db sql.Executor, id []byte) (buf []byte, err e
 }
 
+// NonceByID retrieves VRFNonce corresponding to the specified ATX ID.
+func NonceByID(db sql.Executor, id types.ATXID) (nonce types.VRFPostIndex, err error) {
+	enc := func(stmt *sql.Statement) {
+		stmt.BindBytes(1, id.Bytes())
+	}
+	gotNonce := false
+	dec := func(stmt *sql.Statement) bool {
+		if stmt.ColumnType(0) != sqlite.SQLITE_NULL {
+			nonce = types.VRFPostIndex(stmt.ColumnInt64(0))
+			gotNonce = true
+		}
+		return true
+	}
+
+	if rows, err := db.Exec("select nonce from atxs where id = ?1", enc, dec); err != nil {
+		return types.VRFPostIndex(0), fmt.Errorf("get nonce for ATX id %v: %w", id, err)
+	} else if rows == 0 || !gotNonce {
+		return types.VRFPostIndex(0), sql.ErrNotFound
+	}
+
+	return nonce, err
+}
```
This seems incorrect. The nonce can change over time:

- epoch n: nonce == 10
- epoch n+1: nonce == 5
- epoch n+2: nonce == 7

`NonceByID` is undefined - depending on which epoch you are asking for it might be a different value.

fasmat replied:
Sorry my bad - I misunderstood what the function actually does. Ignore my previous comment 🙂

fasmat commented on sql/atxs/atxs.go:
```diff
@@ -350,8 +359,66 @@ func getBlob(ctx context.Context, db sql.Executor, id []byte) (buf []byte, err e
+		}
+	}
+
+	if err := add(db, atx, atx.VRFNonce); err != nil {
```
Would this function for an ATX with a prev ATX but without a nonce store `NULL` in the database for the nonce? I thought we wanted to have the following logic:

1. ATX has a nonce -> store it with the nonce it contains
2. ATX has no nonce -> look at the prev ATX and store the new ATX with the same nonce as the prev ATX (prev ATX will always have a nonce).
3. an ATX without a prev ATX always has a nonce - no need to check again here (this is already what the handler validates)

ivan4th replied:
I actually tried enforcing the rule of always having the nonce one way or the other (marking the column as `NOT NULL`), but sadly it broke too many tests :( Probably could be done as a followup, as fixing all the tests that don't care about nonces and prev ATXs is quite a bit of work

fasmat replied:
OK, then can you add a GH issue and a TODO in the code so we don't forget to fix that in a future PR? 🙂

ivan4th replied:
Sure, created #5844

fasmat commented on sql/atxs/atxs.go:
```diff
@@ -350,8 +359,66 @@ func getBlob(ctx context.Context, db sql.Executor, id []byte) (buf []byte, err e
+	return atx.VRFNonce, nil
+}
+
+// AddMaybeNoNonce adds an ATX for a given ATX ID. It doesn't try
+// to set the nonce field if VRFNonce is not set in the ATX.
+func AddMaybeNoNonce(db sql.Executor, atx *types.VerifiedActivationTx) error {
+	return add(db, atx, atx.VRFNonce)
+}
```
I'm not understanding this; `Add` and `AddGettingNonce` also both don't store a nonce if the ATX has none?

I thought the goal should be that every ATX that is stored in the DB has a nonce in it's row

ivan4th replied:
This function is only to be used in tests. Adding the comment for it

fasmat commented on sql/atxs/atxs.go:
```diff
@@ -486,20 +571,29 @@ func LatestN(db sql.Executor, n int) ([]CheckpointAtx, error) {
 		stmt.ColumnBytes(5, catx.SmesherID[:])
 		catx.Sequence = uint64(stmt.ColumnInt64(6))
 		stmt.ColumnBytes(7, catx.Coinbase[:])
+		if sql.IsNull(stmt, 8) {
+			ierr = errors.New("missing nonce")
+			return false
+		} else {
+			catx.VRFNonce = types.VRFPostIndex(stmt.ColumnInt64(8))
+		}
```
This will break checkpoint restore - if the ATX was published without a VRF nonce it has to be returned here without one.

Restoring the checkpoint involves running all ATXs of the checkpoint through the ATX handler again. If you changed them (by adding or removing the nonce) they will fail to validate.

Ideally this function should really just read the blobs, decode them and ignore what's in the main atx table.

In a later step we will have to change checkpoint to not even require to decode the ATXs at all.

cc @poszu

ivan4th replied:
`catx.VRFNonce` was being populated in the `checkpointDB` function: https://github.com/spacemeshos/go-spacemesh/pull/5791/files#diff-c6ae423644f8c867db6bf09a7d97c28e23ce4a00f2c2fe7a0697739061fb1561L74-L79
by means of `VRFNonce` function which does epoch+pubkey based query
The `nonce` column here doesn't necessarily correspond to the value of `VRFNonce` field in the table, but rather basically is what `VRFNonce` func would have returned for the atxs' `NodeID` and `Epoch+1`.

fasmat commented on sql/atxs/atxs.go:
```diff
@@ -569,27 +670,46 @@ func IterateAtxsData(
 			ticks := uint64(stmt.ColumnInt64(6))
-			return fn(id, node, epoch, coinbase, effectiveUnits*ticks, baseHeight, baseHeight+ticks)
+			var vrfNonce *types.VRFPostIndex
+			if stmt.ColumnType(7) != sqlite.SQLITE_NULL {
+				nonce := types.VRFPostIndex(stmt.ColumnInt64(7))
+				vrfNonce = &nonce
+			}
```
I thought this column will never be NULL?

ivan4th replied:
As `NOT NULL` rule cannot be enforced yet due to test issues (see above), this makes it easier to check if it was violated (during warmup)

fasmat commented on checkpoint/runner.go:
```diff
@@ -71,12 +71,7 @@ func checkpointDB(
 		if err != nil {
 			return nil, fmt.Errorf("atxs snapshot commitment: %w", err)
 		}
-		vrfNonce, err := atxs.VRFNonce(tx, catx.SmesherID, catx.Epoch+1)
-		if err != nil {
-			return nil, fmt.Errorf("atxs snapshot nonce: %w", err)
-		}
 		copy(atxSnapshot[i].CommitmentATX[:], commitmentAtx[:])
-		atxSnapshot[i].VRFNonce = vrfNonce
```
This breaks checkpoint, I believe.

ivan4th replied:
This value is now obtained by `LatestN` function by just taking the value of the `nonce` column for the ATX. The value of `nonce` column is not the same as `VRFNonce` field stored in the blob, but rather the same nonce that is returned by `atxs.VRFNonce(tx, catx.SmesherID, catx.Epoch+1)` here

fasmat commented on sql/atxs/atxs_test.go:
```diff
@@ -54,7 +54,7 @@ func TestGet(t *testing.T) {
 
 func TestAll(t *testing.T) {
-	db := sql.InMemory()
+	db := sql.InMemory(sql.WithConnections(10000))
```
That seems really high - the default is 16. Does this mean the tested functions now need significantly more connections to be able to fetch the data?

ivan4th replied:
Forgot to remove a temporary setting, removing it now

fasmat commented on sql/migrations.go:
```diff
@@ -73,8 +73,7 @@ func LocalMigrations() ([]Migration, error) {
 
-func sqlMigrations(dbname string) ([]Migration, error) {
-	var migrations []Migration
+func sqlMigrations(dbname string, migrations ...Migration) ([]Migration, error) {
```
Did you have something in mind with the `migrations` parameter? I don't see it being used anywhere?

ivan4th replied:
Forgot to remove, removing it now

ivan4th committed:
sql: add comment for AddMaybeNoNonce

This func is only to be used in tests

ivan4th committed:
Fix minor style issues

ivan4th committed:
sql: simplify migration 17 by removing unneeded atx map

ivan4th committed:
sql: further simplify migration 17

Remove dangling map (only needed for few equivocating ATXs)

poszu commented on sql/atxs/atxs.go:
```diff
@@ -366,28 +434,43 @@ func Add(db sql.Executor, atx *types.VerifiedActivationTx) error {
 
 	_, err = db.Exec(`
 		insert into atxs (id, epoch, effective_num_units, commitment_atx, nonce,
-			 pubkey, atx, received, base_tick_height, tick_count, sequence, coinbase, validity)
-		values (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13);`, enc, nil)
+			 pubkey, received, base_tick_height, tick_count, sequence, coinbase,
+			 validity, prev_id)
+		values (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13)`, enc, nil)
 	if err != nil {
 		return fmt.Errorf("insert ATX ID %v: %w", atx.ID(), err)
 	}
+
+	enc = func(stmt *sql.Statement) {
+		stmt.BindBytes(1, atx.ID().Bytes())
+		stmt.BindBytes(2, buf)
+	}
+	_, err = db.Exec("insert into atx_blobs (id, atx) values (?1, ?2)", enc, nil)
```
It should probably be enforced that these 2 inserts happen atomically. Maybe via taking an `sql.Tx` as the argument?

ivan4th replied:
There are a lot of tests that don't use transactions when creating ATXs and they will have to be modified. The possible non-atomic insertion may only affect the tests, as the actual creation of ATXs always happens within a transaction (in `activation/handler.go`). Maybe something to do in a follow-up along with enforcing `NOT NULL` on the `nonce` field?

ivan4th replied:
Noted the problem as tech debt in #5844

ivan4th committed:
Merge branch 'develop' into fix/quicker-cache-warmup

ivan4th committed:
sql: fix misc script issues

ivan4th commented:
bors merge
